lib/inventory/controller.go

```diff
@@ -51,24 +55,50 @@ const (
 	sshUpsertRetryOk  testEvent = "ssh-upsert-retry-ok"
 	sshUpsertRetryErr testEvent = "ssh-upsert-retry-err"
 
+	instanceHeartbeatOk  testEvent = "instance-heartbeat-ok"
+	instanceHeartbeatErr testEvent = "instance-heartbeat-err"
+
+	instanceCompareFailed testEvent = "instance-compare-failed"
+
 	handlerStart = "handler-start"
 	handlerClose = "handler-close"
 )
 
+// intervalKey is used to uniquely identify the subintervals registerd with the interval.MultiInterval
+// instance that we use for managing periodics associated with upstream handles.
+type intervalKey int
+
+const (
+	instanceHeartbeatKey intervalKey = 1
+	serverKeepAliveKey   intervalKey = 2
+)
+
 type controllerOptions struct {
-	serverKeepAlive  time.Duration
-	testEvents       chan testEvent
-	maxKeepAliveErrs int
+	serverKeepAlive    time.Duration
+	instanceHBInterval time.Duration
+	testEvents         chan testEvent
+	maxKeepAliveErrs   int
+	authID             string
 }
 
 func (options *controllerOptions) SetDefaults() {
+	baseKeepAlive := apidefaults.ServerKeepAliveTTL()
 	if options.serverKeepAlive == 0 {
-		baseKeepAlive := apidefaults.ServerKeepAliveTTL()
 		// use 1.5x standard server keep alive since we use a jitter that
 		// shortens the actual average interval.
 		options.serverKeepAlive = baseKeepAlive + (baseKeepAlive / 2)
 	}
 
+	if options.instanceHBInterval == 0 {
+		// we use 3.5x the server keep alive ttl in order to limit the amount of instance
+		// hbs since they are potentially more costly. 3.5x is arbitrary, but since server expiry
+		// is 10x keepalive TTL under standard configuration, this lets us be certain that we'll
+		// heartbeat at most 3 times within that window, which is a reasonable target
+		// when trying to strike a balance between reliable heartbeating and minimizing the increased
+		// resource utilization incurred by this feature.
+		options.instanceHBInterval = (baseKeepAlive * 3) + (baseKeepAlive / 2)
```

Comment:
I find `baseKeepAlive * 7 / 2` to be confusing.  It would evaluate correctly, but to me it reads a lot like `baseKeepAlive * (7 / 2)`, which would not evaluate correctly.  I usually prefer to make mathematical order of operations explicit, just for my own sanity.